File: app/src/main/java/ru/starksoft/differ/sample/base/divider/DividerPaintItemDecoration.java
package ru.starksoft.differ.sample.base.divider;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.view.View;

import androidx.annotation.ColorRes;
import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

public abstract class DividerPaintItemDecoration extends RecyclerView.ItemDecoration {

    private final Paint paint;

    private int color;

    public DividerPaintItemDecoration(@NonNull Context context, @ColorRes int colorResId, int dividerWidthDimenResId) {
        color = ContextCompat.getColor(context, colorResId);
        paint = new Paint();
        paint.setColor(color);
        paint.setStrokeWidth((int) context.getResources().getDimension(dividerWidthDimenResId));
    }

    @NonNull
    protected abstract ExtraPadding getExtraPadding(@IntRange(from = 0) int childPosition);

    @Override
    public void onDrawOver(@NotNull Canvas c, @NotNull RecyclerView parent, @NotNull RecyclerView.State state) {
        super.onDraw(c, parent, state);

        paint.setColor(color);
        int alpha = paint.getAlpha();

        int childCount = parent.getChildCount();

        for (int i = 0; i < childCount - 1; i++) {
            View child = parent.getChildAt(i);

            int childPosition = parent.getChildAdapterPosition(child);

            if (childPosition >= 0 && disableDraw(childPosition) || childPosition < 0) {
                continue;
            }

            paint.setAlpha((int) (child.getAlpha() * alpha));


            ExtraPadding extraPadding = getExtraPadding(childPosition);
            int startX = parent.getPaddingLeft() + extraPadding.left + (int) child.getTranslationX();
            int stopX = parent.getWidth() - parent.getPaddingRight() - extraPadding.right + (int) child.getTranslationX();

            RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
            int y;
            if (isTopDivider(childPosition)) {
                y = child.getTop() + params.topMargin + (int) child.getTranslationY();
            } else {
                y = child.getBottom() + params.bottomMargin + (int) child.getTranslationY();
            }
            c.drawLine(startX, y, stopX, y, paint);

        }
    }

    public boolean disableDraw(@IntRange(from = 0) int childPosition) {
        return false;
    }

    public boolean isTopDivider(@IntRange(from = 0) int childPosition) {
        return false;
    }

    public static class ExtraPadding {

        public final int left;
        public final int top;
        public final int right;
        public final int bottom;

        public ExtraPadding(int left, int top, int right, int bottom) {
            this.left = left;
            this.top = top;
            this.right = right;
            this.bottom = bottom;
        }
    }
}

File: diff-adapter/src/main/java/ru/starksoft/differ/divider/DividerPaintItemDecoration.kt
package ru.starksoft.differ.divider

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import androidx.annotation.ColorRes
import androidx.annotation.DimenRes
import androidx.annotation.IntRange
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView

abstract class DividerPaintItemDecoration(
	context: Context,
	@ColorRes
	colorResId: Int,
	@DimenRes
	dividerWidthDimenResId: Int
) :
    RecyclerView.ItemDecoration() {

    private val paint = Paint()
    private val color = ContextCompat.getColor(context, colorResId)

    init {
        paint.color = color
        paint.strokeWidth = context.resources.getDimension(dividerWidthDimenResId).toInt().toFloat()
    }

    protected abstract fun getExtraPadding(
		@IntRange(from = 0)
		childPosition: Int
	): ExtraPadding

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDraw(c, parent, state)

        paint.color = color
        val alpha = paint.alpha

        val childCount = parent.childCount

        for (i in 0 until childCount - 1) {
            val child = parent.getChildAt(i)

            val childPosition = parent.getChildAdapterPosition(child)
            if (childPosition >= 0 && disableDraw(childPosition) || childPosition < 0) {
                continue
            }

            paint.alpha = (child.alpha * alpha).toInt()

			val (left, _, right) = getExtraPadding(childPosition)
            val startX = parent.paddingLeft + left + child.translationX.toInt()
            val stopX = parent.width - parent.paddingRight - right + child.translationX.toInt()

            val params = child.layoutParams as RecyclerView.LayoutParams
            val y = if (isTopDivider(childPosition)) {
                child.top + params.topMargin + child.translationY.toInt()
            } else {
                child.bottom + params.bottomMargin + child.translationY.toInt()
            }
            c.drawLine(startX.toFloat(), y.toFloat(), stopX.toFloat(), y.toFloat(), paint)
        }
    }

    open fun disableDraw(
		@IntRange(from = 0)
		childPosition: Int
	): Boolean {
        return false
    }

    open fun isTopDivider(
		@IntRange(from = 0)
		childPosition: Int
	): Boolean {
        return false
    }
}
